### PR TITLE
DEP: Deprecate use of `axis=MAXDIMS` instead of `axis=None`

### PR DESCRIPTION
**Pull request by saswatpp**
addresses #20620 

A deprecation warning has been added whenever `axis==NPY_MAXDIMS` passed in functions. 

@seberg

### CODE REVIEW

File: numpy/core/src/multiarray/conversion_utils.c [hunk 1]
```diff
@@ -271,6 +271,13 @@ PyArray_AxisConverter(PyObject *obj, int *axis)
         if (error_converting(*axis)) {
             return NPY_FAIL;
         }
+        if(*axis == NPY_MAXDIMS){
```
Comment (seberg) [inline numpy/core/src/multiarray/conversion_utils.c:274]:
```suggestion
        if (*axis == NPY_MAXDIMS){
```

File: numpy/core/src/multiarray/conversion_utils.c [hunk 2]
```diff
@@ -271,6 +271,13 @@ PyArray_AxisConverter(PyObject *obj, int *axis)
...
+            /* NumPy 1.23, 2022-01-08 */
+            if (DEPRECATE("axis = 32 is deprecated "
+                          "please use axis = None instead. ") < 0) {
```
Comment (seberg) [inline numpy/core/src/multiarray/conversion_utils.c:277]: I am considering to rephrase this slightly, although it seems fine, hardcoding 32 is also fine (DEPRECATIONS are a bit annoying with formatting).

Maybe:
> Using 32 as an axis is deprecated, it has the same meaning as `None` which should be used instead.  (Deprecated NumPy 1.23)

But not sure. I would like the `(Deprecated NumPy 1.23)` though.

**Review by seberg — COMMENTED** [review]
Sorry for not replying earlier.  This looks great, especially the test!

One last thing we need before putting this in, is that Deprecations should come with a release note.  That means adding a brief file as explained in `docs/release/upcoming_changes/README.rst`.

**Comment by InessaPawson** [comment]
> One last thing we need before putting this in, is that Deprecations should come with a release note. That means adding a brief file as explained in `docs/release/upcoming_changes/README.rst`.

@saswatpp  Let me know if you need help with adding a release note.

File: numpy/core/src/multiarray/conversion_utils.c (hunk 2 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/conversion_utils.c:277]: `np.MAXDIMS` might be better than `32`; or perhaps both.

**Comment by seberg** [comment]
I tweaked the message and added the release note.  Will merge later today if nobody beats me to it, or makes a comment.  Thanks @saswatpp.

**Comment by InessaPawson** [comment]
Hi-five on merging your first pull request to NumPy, @saswatpp!

**Pull request merged**